File: algorithms/easy/101_symmetric_tree.py
# 101. Symmetric Tree

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def is_symmetric(root):
    """Check if a binary tree is symmetric around its center.
    
    Args:
        root: Root of the binary tree
    
    Returns:
        Boolean indicating if the tree is symmetric
    """
    if not root:
        return True
    
    return is_mirror(root.left, root.right)

def is_mirror(left, right):
    """Helper function to check if two subtrees are mirrors of each other.
    
    Args:
        left: Root of left subtree
        right: Root of right subtree
    
    Returns:
        Boolean indicating if the subtrees are mirrors
    """
    # If both nodes are None, they are mirrors
    if not left and not right:
        return True
    
    # If one is None and the other isn't, they are not mirrors
    if not left or not right:
        return False
    
    # Check if the values are the same and if the subtrees are mirrors
    return (left.val == right.val) and \
           is_mirror(left.left, right.right) and \
           is_mirror(left.right, right.left)

# Iterative implementation using a queue
def is_symmetric_iterative(root):
    if not root:
        return True
    
    from collections import deque
    queue = deque([(root.left, root.right)])
    
    while queue:
        left, right = queue.popleft()
        
        # If both are None, continue
        if not left and not right:
            continue
        
        # If one is None or values are different, return False
        if not left or not right or left.val != right.val:
            return False
        
        # Add the pairs to check
        queue.append((left.left, right.right))
        queue.append((left.right, right.left))
    
    return True

# Test cases
# Example 1: [1,2,2,3,4,4,3] => true
root1 = TreeNode(1)
root1.left = TreeNode(2)
root1.right = TreeNode(2)
root1.left.left = TreeNode(3)
root1.left.right = TreeNode(4)
root1.right.left = TreeNode(4)
root1.right.right = TreeNode(3)
print(is_symmetric(root1))  # True

# Example 2: [1,2,2,null,3,null,3] => false
root2 = TreeNode(1)
root2.left = TreeNode(2)
root2.right = TreeNode(2)
root2.left.right = TreeNode(3)
root2.right.right = TreeNode(3)
print(is_symmetric(root2))  # False

File: algorithms/easy/101_symmetric_tree.js
/**
 * 101. Symmetric Tree
 * 
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function(root) {
    if (!root) return true;
    
    return isMirror(root.left, root.right);
};

/**
 * Helper function to check if two subtrees are mirrors of each other.
 * @param {TreeNode} left - Root of left subtree
 * @param {TreeNode} right - Root of right subtree
 * @return {boolean}
 */
function isMirror(left, right) {
    // If both nodes are null, they are mirrors
    if (!left && !right) return true;
    
    // If one is null and the other isn't, they are not mirrors
    if (!left || !right) return false;
    
    // Check if the values are the same and if the subtrees are mirrors
    return (left.val === right.val) && 
           isMirror(left.left, right.right) && 
           isMirror(left.right, right.left);
}

// Iterative implementation using a queue
var isSymmetricIterative = function(root) {
    if (!root) return true;
    
    const queue = [[root.left, root.right]];
    
    while (queue.length > 0) {
        const [left, right] = queue.shift();
        
        // If both are null, continue
        if (!left && !right) continue;
        
        // If one is null or values are different, return false
        if (!left || !right || left.val !== right.val) return false;
        
        // Add the pairs to check
        queue.push([left.left, right.right]);
        queue.push([left.right, right.left]);
    }
    
    return true;
};

// Test cases
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val);
    this.left = (left === undefined ? null : left);
    this.right = (right === undefined ? null : right);
}

// Example 1: [1,2,2,3,4,4,3] => true
const root1 = new TreeNode(1);
root1.left = new TreeNode(2);
root1.right = new TreeNode(2);
root1.left.left = new TreeNode(3);
root1.left.right = new TreeNode(4);
root1.right.left = new TreeNode(4);
root1.right.right = new TreeNode(3);
console.log(isSymmetric(root1));  // true

// Example 2: [1,2,2,null,3,null,3] => false
const root2 = new TreeNode(1);
root2.left = new TreeNode(2);
root2.right = new TreeNode(2);
root2.left.right = new TreeNode(3);
root2.right.right = new TreeNode(3);
console.log(isSymmetric(root2));  // false